app/src/main/java/org/jellyfin/androidtv/preference/PreferenceStore.kt

```diff
@@ -1,27 +1,82 @@
+ * different backing stores.
  */
-interface PreferenceStore {
+abstract class PreferenceStore {
 	// val value = store[Preference.x]
-	operator fun <T : Preference<V>, V : Any> get(preference: T): V
-	operator fun <T : Preference<V>, V : Enum<V>> get(preference: T): V
+	@Suppress("UNCHECKED_CAST")
+	operator fun <T : Any> get(preference: Preference<T>): T =
+		// Coerce returned type based on the default type
+		when (preference.defaultValue) {
+			is Int -> getInt(preference.key, preference.defaultValue)
+			is Long -> getLong(preference.key, preference.defaultValue)
+			is Boolean -> getBool(preference.key, preference.defaultValue)
+			is String -> getString(preference.key, preference.defaultValue)
+			else -> throw IllegalArgumentException("${preference.type.simpleName} type is not supported")
+		} as T
+
+	// val value = store[Preference.x]
+	operator fun <T : Enum<T>> get(preference: Preference<T>) = getEnum(preference)
 
 	// store[Preference.x] = value
-	operator fun <T : Preference<V>, V : Any> set(preference: T, value: V)
-	operator fun <T : Preference<V>, V : Enum<V>> set(preference: T, value: V)
+	operator fun <T : Any> set(preference: Preference<T>, value: T) =
+		when (value) {
+			is Int -> setInt(preference.key, value)
+			is Long -> setLong(preference.key, value)
+			is Boolean -> setBool(preference.key, value)
+			is String -> setString(preference.key, value)
+			is Enum<*> -> setEnum(preference, value)
+			else -> throw IllegalArgumentException("${preference.type.simpleName} type is not supported")
+		}
 
 	// store.getDefaultValue(Preference.x)
-	fun <T : Preference<V>, V : Any> getDefaultValue(preference: T): V {
+	fun <T : Any> getDefaultValue(preference: Preference<T>): T {
 		return preference.defaultValue
 	}
 
 	// store.reset(Preference.x)
-	fun <T : Preference<V>, V : Any> reset(preference: T) {
+	fun <T : Any> reset(preference: Preference<T>) {
 		this[preference] = getDefaultValue(preference)
 	}
 
 	// store.delete(Preference.x)
-	fun <T : Preference<V>, V : Any> delete(preference: T)
+	abstract fun <T : Any> delete(preference: Preference<T>)
+
+	// Protected methods to get / set items, this is an implementation detail so we protect
+	// it in the abstract common functionality (where it is used)
+	protected abstract fun getInt(keyName: String, defaultValue: Int): Int
+	protected abstract fun getLong(keyName: String, defaultValue: Long): Long
+	protected abstract fun getBool(keyName: String, defaultValue: Boolean): Boolean
+	protected abstract fun getString(keyName: String, defaultValue: String): String
+
+	protected abstract fun setInt(keyName: String, value: Int)
+	protected abstract fun setLong(keyName: String, value: Long)
+	protected abstract fun setBool(keyName: String, value: Boolean)
+	protected abstract fun setString(keyName: String, value: String)
+
+	// Private Enum handling, all Enum types are serialized to / from String types
+	private fun <T : Enum<T>> getEnum(preference: Preference<T>): T {
```

Comment:
:+1: 
Could we move it to a Util class like `PrefEnumSerializer` (or something better named)?

I agree about not having everyone try to serialize enums to string by default, but also would like to avoid duplicating complex code like this into different stores as we have to then maintain n copies of this code.

Reply:
I don't think that's necessary. It's only 3 lines (4 when including the part to get the string value) of code that work, and I don't expect they need to be maintained ever.